### PR TITLE
[Fix] Fix piecewise medium edge case and enable use of LLVM.

### PR DESCRIPTION
**Pull request by frettini**
## Description

This PR introduces 2 fixes to the piecewise volume and integrator:
- Fix piecewise medium and integrator bug where transmittance was set to 0 when pdf was 0. The pdf is not relevant when computing the transmittance analytically so there no need to include it.
- Enable LLVM compatibility with the piecewise medium. This required a refactoring of how the cumulative optical thickness is stored in memory. It now is stored in a dynamic 1D `Float` array  (`FloatStorage`), and  `dr::gather` is leveraged to automatically retrieve an `UnpolarizedSpectrum` from the buffer. Other changes include the use of `dr::eq` instead of `==` and a change in how sigma_t is retrieved from the medium (used to return 0. for llvm).

## Testing

The changes were tested upon the notebooks initially used to develop the piecewise medium which let me compare the performance against a majorant medium. The test suite was also run succesfully.

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `cuda_*` and `llvm_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

### CODE REVIEW

File: src/eradiate_plugins/integrators/piecewise_volpath.cpp [hunk 1]
```diff
@@ -235,7 +230,7 @@ class PiecewiseVolumetricPathIntegrator
                         mei, scene, sampler, medium, channel, active_e);
                     auto [phase_val, phase_pdf] =
                         phase->eval_pdf(phase_ctx, mei, ds.d, active_e);
-                    dr::masked(result, active_e) +=
+                    dr::masked(result, active_e) +=
```
Comment (leroyvn) [inline src/eradiate_plugins/integrators/piecewise_volpath.cpp:233]: Nitpicking:
```suggestion
                    dr::masked(result, active_e) +=
```

File: src/eradiate_plugins/media/piecewise.cpp [hunk 2]
```diff
@@ -264,18 +262,14 @@ class PiecewiseMedium final : public Medium<Float, Spectrum> {
...
-                                  channel));
+            extract_channel(dr::gather<UnpolarizedSpectrum>(m_cum_opt_thickness, opt_start_idx, active), channel),
+            extract_channel(dr::gather<UnpolarizedSpectrum>(m_reverse_cum_opt_thickness, opt_start_idx, active),channel));
         opt_thick_offset -= start_height * sigma_t;
 
         Float log_sample = dr::log(1.f - sample);
         Mask sampled     = dr::zeros<Mask>();
         Mask search      = active && !same_cell;
-
+
```
Comment (leroyvn) [inline src/eradiate_plugins/media/piecewise.cpp:272]: Let's remove the extra spaces:
```suggestion

```

File: src/eradiate_plugins/media/piecewise.cpp [hunk 3]
```diff
@@ -488,24 +478,38 @@ class PiecewiseMedium final : public Medium<Float, Spectrum> {
             std::tie(mei.sigma_s, mei.sigma_n, mei.sigma_t) =
                 get_scattering_coefficients(mei, true);
 
-            cumulative += mei.sigma_t;
-            cum_opt_thickness[i] = cumulative;
+            for(size_t k = 0 ; k<SpectralSize; ++k){
```
Comment (leroyvn) [inline src/eradiate_plugins/media/piecewise.cpp:481]:
```suggestion
            for (size_t k = 0 ; k < SpectralSize; ++k) {
```

File: src/eradiate_plugins/media/piecewise.cpp [hunk 4]
```diff
@@ -488,24 +478,38 @@ class PiecewiseMedium final : public Medium<Float, Spectrum> {
...
+        // Calculate cumulative optical thickness from top to bottom
         for (int32_t i = (int32_t) resolution.z() - 1; i >= 0; --i) {
             mei.p = min + voxel_size * 0.5 + (ScalarFloat) i * step;
             std::tie(mei.sigma_s, mei.sigma_n, mei.sigma_t) =
                 get_scattering_coefficients(mei, true);
 
-            reverse_cumulative += mei.sigma_t;
-            reverse_cum_opt_thickness[(int32_t) resolution.z() - 1 - i] =
-                reverse_cumulative;
+            for(size_t k = 0 ; k<SpectralSize; ++k){
```
Comment (leroyvn) [inline src/eradiate_plugins/media/piecewise.cpp:497]:
```suggestion
            for (size_t k = 0 ; k < SpectralSize; ++k) {
```

File: src/eradiate_plugins/media/piecewise.cpp [hunk 5]
```diff
@@ -488,24 +478,38 @@ class PiecewiseMedium final : public Medium<Float, Spectrum> {
...
+                if constexpr (dr::is_jit_v<UnpolarizedSpectrum>)
+                    reverse_cumulative[k] += ScalarFloat(mei.sigma_t[k][0]);
+                else
+                    reverse_cumulative[k] += ScalarFloat(mei.sigma_t[k]);
+
+                // populate array from 0 to (resolution-1)*SpectralSize
+                reverse_cum_opt_thickness[((int32_t) resolution.z() - 1 - i)*SpectralSize + k] =
```
Comment (leroyvn) [inline src/eradiate_plugins/media/piecewise.cpp:504]:
```suggestion
                reverse_cum_opt_thickness[((int32_t) resolution.z() - 1 - i) * SpectralSize + k] =
```

**Review by leroyvn — COMMENTED** [review]
Hi @frettini, thanks! No comment, only minor formatting remarks. I must say our clang-format setup doesn't handle nested function calls very nicely :(

**Comment by frettini** [comment]
Thanks for the review, I will update the formatting, I was reluctant to run clang because it made the code rather unreadable.

**Pull request merged**